dbms/src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -1755,20 +1787,16 @@ void StorageReplicatedMergeTree::mergeSelectingThread()
 
         try
         {
-            if (need_pull)
-            {
-                /// You need to load new entries into the queue before you select parts to merge.
-                ///  (so we know which parts are already going to be merged).
-                pullLogsToQueue();
-                need_pull = false;
-            }
-
             std::lock_guard<std::mutex> merge_selecting_lock(merge_selecting_mutex);
 
-            /** If many merges is already queued, then will queue only small enough merges.
-              * Otherwise merge queue could be filled with only large merges,
-              *  and in the same time, many small parts could be created and won't be merged.
-              */
+            /// You need to load new entries into the queue before you select parts to merge.
+            ///  (so we know which parts are already going to be merged).
+            /// We must select parts for merge under the mutex because other threads (OPTIMIZE queries) could push new merges.
+            pullLogsToQueue();
```

Comment:
Now we make pull on each iteration.
Previously we did that only after successful merges.
It could increase ZK load significantly.